scheduler/scheduler-examples/src/main/java/org/ow2/proactive/scheduler/examples/SFTPConnector.java

```diff
@@ -0,0 +1,366 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import org.ow2.proactive.scheduler.common.task.TaskResult;
+import org.ow2.proactive.scheduler.common.task.executable.JavaExecutable;
+import org.zeroturnaround.zip.ZipUtil;
+
+import com.jcraft.jsch.Channel;
+import com.jcraft.jsch.ChannelSftp;
+import com.jcraft.jsch.JSch;
+import com.jcraft.jsch.JSchException;
+import com.jcraft.jsch.Session;
+import com.jcraft.jsch.SftpException;
+
+
+/**
+ * Import and export file(s) using SFTPConnector server.
+ *
+ * @author The ProActive Team
+ */
+public class SFTPConnector extends JavaExecutable {
+
+    private static final String DEFAULT_SFTP_HOSTNAME = "localhost";
+
+    private static final int DEFAULT_SFTP_PORT = 22;
+
+    private String sftpHostname = DEFAULT_SFTP_HOSTNAME;
+
+    private int sftpPort = DEFAULT_SFTP_PORT;
+
+    private boolean sftpExtractArchive;
+
+    private String sftpLocalRelativePath = null;
+
+    private String sftpRemoteRelativePath = null;
+
+    private String sftpMode = null;
+
+    private String sftpUsername = null;
+
+    private String sftpPassword = null;
+
+    private static final String SFTP_LOCAL_RELATIVE_PATH_ARG = "sftpLocalRelativePath";
+
+    @Override
+    public void init(Map<String, Serializable> args) throws Exception {
+        if (args.containsKey("sftpHostname")) {
+            sftpHostname = args.get("sftpHostname").toString();
+        }
+        if (args.containsKey("sftpPort")) {
+            sftpPort = Integer.parseInt(args.get("sftpPort").toString());
+        }
+        if (args.containsKey("sftpMode")) {
+            sftpMode = args.get("sftpMode").toString();
+        }
+        if (args.containsKey("sftpExtractArchive")) {
+            sftpExtractArchive = Boolean.parseBoolean(args.get("sftpExtractArchive").toString());
+        }
+        if (args.containsKey(SFTP_LOCAL_RELATIVE_PATH_ARG) &&
+            !args.get(SFTP_LOCAL_RELATIVE_PATH_ARG).toString().isEmpty()) {
+            sftpLocalRelativePath = args.get(SFTP_LOCAL_RELATIVE_PATH_ARG).toString();
+        } else {
+            //we throw an exception tp prevent transferring all the contents of the global space.
+            if (sftpMode.equals("PUT")) {
+                throw new IllegalArgumentException("you have to specify the local relative path. empty value is not allowed.");
+            }
+            //default value is getlocalspace() because it will always be writable and moreover can be used to transfer files to another data space(global, user)
+            sftpLocalRelativePath = getLocalSpace();
+        }
+        if (args.containsKey("sftpRemoteRelativePath")) {
+            sftpRemoteRelativePath = args.get("sftpRemoteRelativePath").toString();
+        }
+
+        sftpUsername = getThirdPartyCredential("SFTP_USERNAME");
+        sftpPassword = getThirdPartyCredential("SFTP_PASSWORD");
```

Comment:
I am thinking that maybe there is a better way to store username/password combination. Currently, it is only possible to interact with one sftp server for a given scheduler user.

The user name could be passed as a task argument, and the password could be stored as:
key:
sftp://sftpHostname:sftpPort 
value:
password

The same comment applies for the FTPConnector

Reply:
sorry, I meant:
key:
username@sftp://sftpHostname:sftpPort

Reply:
This is a good point. In fact, we have the same problem for all data connectors (you can have only one connection per workflow unless you manually add prefixes to credentials variables as well as to Hostname, etc...) This would be a great [temporary?] fix for all data-connectors.

Btw, this is not related, but we kind of (KIND OF) serve the same purpose with the URL-connector task wich in turn calls the sftp workflow from the catalog after extracting all the stuff from the url. However, we don't properly deal yet with the password, but at least you can have as many sftp/ftp connections you want with URL tasks without risking any collision.